homework/index.js

```diff
@@ -0,0 +1,144 @@
+'use strict';
+
+function main() {
+
+    const mainUrl = 'https://api.github.com/orgs/HackYourFuture/repos?per_page=100';
+
+    // initial page: create header and main container.
+    const root = document.getElementById('root');
+    const header = createAndAppend('header', root);
+    createAndAppend('h3', header, { text: 'HYF Repositories' });
+    const mainContainer = createAndAppend('main', root, { class: 'main-container' });
+
+    fetchAndRender(mainUrl);
+
+    async function fetchAndRender(url) {
+
+        try {
+            const data = await fetchJSON(url);
+            const rep = await renderReposList(data);
+            const contributors = await fetchJSON(rep.contributors_url);
+            renderInfo(rep);
+            renderContributions(contributors);
+        } catch (err) {
+            renderError(err);
+        }
+
+    }
+
+    function fetchJSON(url) {
+
+        return new Promise((resolve, reject) => {
+            const XHR = new XMLHttpRequest();
+            XHR.open('GET', url, true);
+            XHR.responseType = 'json';
+            XHR.onload = () => {
+                if (XHR.status < 400) {
+                    // resolve the promise if the request is OK
+                    resolve(XHR.response);
+
+                } else {
+                    // if bad response reject the promise with error object
+                    reject(new Error(`Network error: ${XHR.status} - ${XHR.statusText}`));
+
+                }
+            };
+            //if XHR not loaded reject the promise
+            XHR.onerror = () => reject(new Error('Network request failed'));
+            XHR.send();
+        });
+
+    }
+
+    async function renderReposList(reposObj) {
+
+        const header = document.querySelector('header');
+        reposObj.sort((a, b) => { return a.name.localeCompare(b.name); });
+        //create list and append it to header
+        const selectList = createAndAppend('select', header, { id: 'selectList' });
+
+        for (const rep in reposObj) {
+            // add an option for each repo with value that is his index at reposObject 
+            createAndAppend('option', selectList, { value: rep, text: reposObj[rep].name });
+
+        }
+        //add a listener when select a new option
+        selectList.onchange = async () => {
+
+            try {
+                const contributors = await fetchJSON(reposObj[selectList.value].contributors_url);
+                renderInfo(reposObj[selectList.value]);
+                renderContributions(contributors);
+            } catch (err) {
+                renderError(err);
+            }
+
+        };
+        return reposObj[selectList.value];
+
+    }
+
```

Comment:
I think you could further improve your code by reorganizing some functions and introduce a new function to render a given repo (remove blank lines and comments for brevity):

```js
async function fetchAndRender(url) {
  try {
    const data = await fetchJSON(url);
    renderReposList(data);
  } catch (err) {
    renderError(err);
  }
}

async function renderReposList(repos) {
  const header = document.querySelector('header');
  repos.sort((a, b) => { return a.name.localeCompare(b.name); });
  const selectList = createAndAppend('select', header, { id: 'selectList' });
  repos.forEach((repo, index) => {
    createAndAppend('option', selectList, { value: index, text: repo.name });
  });
  selectList.onchange = async () => renderRepo(repos[selectList.value]);
  // show first repo at startup
  renderRepo(repos[selectList.value]);
}

async function renderRepo(repo) {
  try {
    const contributors = await fetchJSON(repo.contributors_url);
    renderInfo(repo);
    renderContributions(contributors);
  } catch (err) {
    renderError(err);
  }
}
```